core/templates/pages/topics-and-skills-dashboard-page/modals/unassign-skill-from-topics-modal.component.ts

```diff
@@ -42,52 +43,111 @@ export class UnassignSkillFromTopicsModalComponent
   // and we need to do non-null assertion. For more information, see
   // https://github.com/oppia/oppia/wiki/Guide-on-defining-types#ts-7-1
   skillId!: string;
-  topicsAssignments!: TopicAssignments;
+  eligibleTopicNameToTopicAssignments!: TopicNameToTopicAssignments;
+  nonEligibleTopicNameToTopicAssignments !: TopicNameToTopicAssignments;
+  nonEligibleTopicNames: string[] = [];
   topicsAssignmentsAreFetched: boolean = false;
   selectedTopicNames: string[] = [];
   selectedTopics: TopicAssignmentsSummary[] = [];
+  eligibleTopicNamesForUnassignment: string[] = [];
 
   constructor(
     private ngbActiveModal: NgbActiveModal,
     private topicsAndSkillsDashboardBackendApiService:
-    TopicsAndSkillsDashboardBackendApiService
+    TopicsAndSkillsDashboardBackendApiService,
+    private urlInterpolationService: UrlInterpolationService
   ) {
     super(ngbActiveModal);
   }
 
+  fetchTopicIdToDiagnosticTestSkillIds(
+      topicAssignments: AssignedSkill[]): void {
+    let allTopicIds = [];
+    for (let topic of topicAssignments) {
+      allTopicIds.push(topic.topicId);
+    }
+
+    this.eligibleTopicNameToTopicAssignments = {};
+    this.nonEligibleTopicNameToTopicAssignments = {};
+    let eligibleTopicIds: string[] = [];
+    let nonEligibleTopicIds: string[] = [];
+    this.topicsAndSkillsDashboardBackendApiService
+      .fetchTopicIdToDiagnosticTestSkillIdsAsync(allTopicIds).then(
+        (responseDict: TopicIdToDiagnosticTestSkillIdsResponse) => {
+          for (let topicId in responseDict.topicIdToDiagnosticTestSkillIds) {
+            let diagnosticTestSkillIds = (
+              responseDict.topicIdToDiagnosticTestSkillIds[topicId]);
+
+            if (
+              diagnosticTestSkillIds.length === 1 &&
+              diagnosticTestSkillIds.indexOf(this.skillId) !== -1
+            ) {
+              nonEligibleTopicIds.push(topicId);
+            } else {
+              eligibleTopicIds.push(topicId);
+            }
+          }
+
+          for (let topic of topicAssignments) {
+            if (nonEligibleTopicIds.indexOf(topic.topicId) !== -1) {
```

Comment:
Choose the dict to add the topic dict to based on the eligibility, and store a reference to that in a variable. Then add the subtopicId/topicVersion/topicId dict to that.

Then you won't need the if/else and the code will be shorter/cleaner.

Reply:
Thank you for the suggestion, I have refactored the logic. It seems pretty clear now. Please provide your feedback.